src/realm/util/compression.cpp

```diff
@@ -0,0 +1,942 @@
+        // feeding in more input
+        while (rc != COMPRESSION_STATUS_END) {
+            auto dst_ptr_start = strm.dst_ptr;
+            rc = compression_stream_process(&strm, 0);
+            if (rc == COMPRESSION_STATUS_ERROR)
+                return error::corrupt_input;
+            if (strm.dst_ptr == dst_ptr_start)
+                break;
+
+            // libcompression doesn't check the checksum, so do it manually.
+            // This loop will never actually run multiple times as in practice
+            // libcompression doesn't actually write more bytes than fit in uLong
+            // in a single call to compression_stream_process()
+            while (dst_ptr_start < strm.dst_ptr) {
+                auto size = bounded_avail(strm.dst_ptr - dst_ptr_start);
+                actual_checksum = adler32(actual_checksum, dst_ptr_start, size);
+                dst_ptr_start += size;
+            }
+        }
+
+        // The checksum at the end can potentially be straddling a block boundary
+        // and we can't rewind, so maintain a rolling window of the last four
+        // bytes seen.
+        for (uint8_t byte : compressed_buf.last(std::min<size_t>(4u, compressed_buf.size()))) {
+            expected_checksum <<= 8;
+            expected_checksum += byte;
+        }
+    } while ((compressed_buf = compressed.next_block()), compressed_buf.size());
+    rc = compression_stream_process(&strm, COMPRESSION_STREAM_FINALIZE);
+    if (rc != COMPRESSION_STATUS_END)
+        return error::corrupt_input;
+    if (strm.dst_size != 0)
+        return error::incorrect_decompressed_size;
+    if (expected_checksum != actual_checksum)
+        return error::corrupt_input;
+    // Check for remaining extra input
+    if (strm.src_size || compressed.next_block().size())
+        return error::corrupt_input;
+    return std::error_code{};
+}
+#endif
+
+std::error_code decompress(NoCopyInputStream& compressed, Span<const char> compressed_buf,
+                           Span<char> decompressed_buf, Algorithm algorithm, bool has_header)
+{
+    using namespace compression;
+
+    if (decompressed_buf.size() == 0) {
+        return std::error_code{};
+    }
+    if (!compressed_buf.size()) {
+        return error::incorrect_decompressed_size;
+    }
+
+#if REALM_USE_LIBCOMPRESSION
+    // All of our non-macOS deployment targets are high enough to have libcompression,
+    // but we support some older macOS versions
+    if (__builtin_available(macOS 10.11, *)) {
+        if (algorithm != Algorithm::None)
+            return decompress_libcompression(compressed, compressed_buf, decompressed_buf, algorithm, has_header);
+    }
+#endif
+
+    switch (algorithm) {
+        case Algorithm::None:
+            return decompress_none(compressed, compressed_buf, decompressed_buf);
+        case Algorithm::Deflate:
+            return decompress_zlib(compressed, compressed_buf, decompressed_buf, has_header);
+        default:
```

Comment:
perhaps handling Algorithm::Lzfse here and returning a "not supported on this platform" would be more user friendly than a generic corruption message just in case someone is tries to copy a sync'd Realm from Apple to Android